File: api/models/queue.py
"""
Pydantic models for queue items.

These define what a valid queue item looks like — both what comes IN from
the React form (QueueItemCreate) and what goes OUT to the table (QueueItem).

The split between Create and full Item matters because:
- When creating, the user provides part_number, process, spec, etc.
- The API fills in submitted_by from the JWT (not user-supplied)
- Fields like "sent" are system-managed, not user-supplied
"""

from pydantic import BaseModel
from typing import Optional


class QueueItemCreate(BaseModel):
    part_number: str
    process: str
    spec: str
    material: Optional[str] = ""
    quantities: Optional[str] = ""
    due_date: Optional[str] = None
    qt_so_number: Optional[str] = ""   # maps to "qt/so #" column in the CSV
    cui_itar: Optional[bool] = False
    rev: Optional[str] = ""
    notes: Optional[str] = ""


class QueueItem(QueueItemCreate):
    # These fields exist in the CSV but aren't set by the user on creation
    sent: Optional[str] = ""
    submitted_by: Optional[str] = ""
    file_location: Optional[str] = ""
    print_callout: Optional[str] = ""

    # from_attributes=True lets Pydantic build this model from a dict
    # (pandas .to_dict("records") produces dicts, not QueueItem objects)
    model_config = {"from_attributes": True}

File: frontend/src/api/queue.ts
import client from './client'

export interface QueueItem {
  part_number: string
  process: string
  spec: string
  material?: string
  quantities?: string
  due_date?: string
  qt_so_number?: string
  cui_itar?: boolean
  rev?: string
  notes?: string
  sent?: string
  submitted_by?: string
}

// Omit<> is a TypeScript utility that creates a new type by removing fields.
// When creating a new item, the server fills in sent and submitted_by —
// the user doesn't supply them. This prevents the form from even having those fields.
export type QueueItemCreate = Omit<QueueItem, 'sent' | 'submitted_by'>

export async function fetchQueue(): Promise<QueueItem[]> {
  const { data } = await client.get<QueueItem[]>('/queue/')
  return data
}

export async function addQueueItem(item: QueueItemCreate): Promise<QueueItem> {
  const { data } = await client.post<QueueItem>('/queue/', item)
  return data
}

export async function removeQueueItem(partNumber: string): Promise<void> {
  await client.delete(`/queue/${encodeURIComponent(partNumber)}`)
}

export interface QueueItemUpdate {
  current_process: string
  spec?: string
  material?: string
  quantities?: string
  qt_so_number?: string
  cui_itar?: string
  rev?: string
  notes?: string
  due_date?: string
  file_location?: string
}

export async function updateQueueItem(partNumber: string, body: QueueItemUpdate): Promise<QueueItem> {
  const { data } = await client.patch<QueueItem>(`/queue/${encodeURIComponent(partNumber)}`, body)
  return data
}

export async function fetchProcesses(): Promise<string[]> {
  const { data } = await client.get<string[]>('/specs/processes')
  return data
}

export async function fetchSpecsForProcess(process: string): Promise<string[]> {
  const { data } = await client.get<string[]>(`/specs/processes/${encodeURIComponent(process)}/specs`)
  return data
}
